applications/AdjointFluidApplication/custom_processes/input_primal_solution_process.h

```diff
@@ -167,6 +185,122 @@ class InputPrimalSolutionProcess : public Process
 
     void ExecuteBeforeSolutionLoop() override
     {
+        KRATOS_TRY
+
+        // read the values from the file before starting the computation
+        // open file to read data
+        H5::H5File File(mFilename.c_str(), H5F_ACC_RDONLY);
+        
+        H5::DataSet NodeIdDataset = File.openDataSet("/NodalData/Id");
+
+        hsize_t dims[3];
+        NodeIdDataset.getSpace().getSimpleExtentDims(dims);
+
+        if (mNumNodes != dims[0])
+            KRATOS_THROW_ERROR(std::runtime_error, "inconsistent dimension in file: ", mFilename)
+
+        if (mNumNodes != static_cast<SizeType>(mrModelPart.Nodes().size()))
+            KRATOS_THROW_ERROR(std::runtime_error, "detected change in number of nodes.", "")
+
+        std::vector<unsigned int> NodeIdBuffer(mrModelPart.Nodes().size());
+        NodeIdDataset.read(NodeIdBuffer.data(), H5::PredType::NATIVE_UINT);
+
+        // check that the node ordering is consistent between the model part and file
+        int Delta = 0;
+        auto itNode = std::begin(NodeIdBuffer);
+        std::for_each(std::begin(mrModelPart.Nodes()), std::end(mrModelPart.Nodes()),
+            [&](NodeType& rNode)
+            {
+                Delta += std::abs(static_cast<int>(rNode.Id()) - static_cast<int>(*itNode));
+                itNode++;
```

Comment:
Casts again. Also, I know I am the first one who did this, but it is more clear to use the range-based for in such cases
for (const auto& rNode : mrModelPart.Nodes()), than the std::for_each.